File: Codice/sgad/servertier/src/main/scala/sgad/servertier/dataaccess/data/shareddata/Unit.scala
/**
 * FILE: Unit.scala
 * PERCORSO /Codice/sgad/servertier/src/main/scala/sgad/servertier/dataaccess/data/shareddata
 * DATA CREAZIONE: 18 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-18 - Creazione della classe - Segantin Fabio
 */
package sgad.servertier.dataaccess.data.shareddata

/**
 * Classe per la gestione delle informazioni di una unità.
 * @constructor
 * @param name Il nome dell'unità.
 * @param attack La forza di attacco dell'unità.
 * @param defence La forza di difesa dell'unità.
 * @param cost Il costo di produzione dell'unità.
 * @param isBuilder Identifica le unità che permettono di costruire.
 */
class Unit(private val name: String, private val attack: Int, private val defence: Int, private val cost: Cost, private val isBuilder: Boolean) {
	if (name == null
		|| cost == null
		|| attack < 0
		|| defence < 0)
		throw new IllegalArgumentException

	/**
	 * Metodo getter per l'attributo name.
	 * @return il nome dell'unità.
	 */
	def getName = name

	/**
	 * Metodo getter per l'attributo attack.
	 * @return la forza di attacco dell'unità.
	 */
	def getAttack = attack

	/**
	 * Metodo getter per l'attributo defence.
	 * @return la forza di difesa dell'unità.
	 */
	def getDefence = defence

	/**
	 * Metodo getter per l'attributo cost.
	 * @return il costo di produzione dell'unità.
	 */
	def getCost = cost

	/**
	 * Fornisce la chiave identificativa dell'unità.
	 * @return La stringa chiave dell'unità.
	 */
	def getKey = getName

	/**
	 * Metodo getter per l'attributo isBuilder.
	 * @return La variabile isBuilder.
	 */
	def getIsBuilder = isBuilder

	/**
	 * Metodo per controllare l'uguaglianza.
	 * @param other L'altro oggetto con cui si fa il confronto.
	 * @return True o false in seguito al confronto.
	 */
	override def equals(other: Any): Boolean = {
		if (!other.isInstanceOf[Unit])
			false
		else
			getKey == other.asInstanceOf[`Unit`].getKey
	}
}

File: Codice/sgad/clienttier/model/generaldata/Unit.js
/**
 * FILE: Unit.js
 * PERCORSO /Codice/sgad/clienttier/model/generaldata/Unit.js
 * DATA CREAZIONE: 14 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-15 - Implementazione dei metodi getKey(), getName(), getAttack(), getDefence(), getCost() e valueOf() - Gallo Francesco
 * 2014-02-14 - Creazione della classe - Gallo Francesco
 */
 
 /**
 * Classe per la gestione delle informazioni di una unità.
 * @constructor
 * @param {string} name Il nome dell'unità.
 * @param {int} attack La forza di attacco dell'unità.
 * @param {int} defence La forza di difesa dell'unità.
 * @param {Cost} cost Il costo di produzione dell'unità.
 */
function Unit(name, attack, defence, cost) {
	/**
	 * Il nome dell'unità.
	 * @type {string}
	 * @private
	 */
	this.name = name;
	/**
	 * La forza di attacco dell'unità.
	 * @type {int}
	 * @private
	 */
	this.attack = attack;
	/**
	 * La forza di difesa dell'unit?.
	 * @type {int}
	 * @private
	 */
	this.defence = defence;
	/**
	 * Il costo di produzione dell'unit?.
	 * @type {Cost}
	 * @private
	 */
	this.cost = cost;
}

/**
 * Fornisce la chiave identificativa dell'unità.
 * @returns {string} La chiave identificativa.
 */
Unit.prototype.getKey = function() {
    return this.getName();
};

/**
 * Metodo getter del nome dell'unità.
 * @returns {string} Il nome dell'unità.
 */
Unit.prototype.getName = function() {
    return this.name;
};

/**
 * Metodo getter del valore di attacco dell'unità.
 * @returns {int} Il valore di attacco dell'unità.
 */
Unit.prototype.getAttack = function() {
    return this.attack;
};

/**
 * Metodo getter del valore di difesa dell'unità.
 * @returns {int} Il valore di difesa dell'unità.
 */
Unit.prototype.getDefence = function() {
    return this.defence;
};

/**
 * Metodo getter del costo dell'unità.
 * @returns {Cost} Il costo dell'unità.
 */
Unit.prototype.getCost = function() {
    return this.cost;
};
/**
 * Ridefinizione del metodo valueOf di Unit.
 * @override
 * @returns {string} Il nome identificativo dell'unità.
 */
Unit.prototype.valueOf = function() {
    return  JSON.stringify({name : this.getKey()});
};